### PR TITLE
SRV_Channel: Passthrough mapped RC channels

### PR DESCRIPTION
**Pull request by TunaLobster**
The intention of this is to create new servo functions that map the RC input to the servo min/trim/max. I think the function I am needing already exists to map between the 2. Just getting lost in calls right now. I'll take another read through the code to figure it out.

### CODE REVIEW

**Comment by Hwurzburg** [comment]
I think a bitmask to determine if the aux function 51-66 is a direct passthru like now, or a output scaled using the output min/max/trim, would be a better approach than new functions

**Comment by timtuxworth** [comment]
Maybe I'm missing something, but doesn't setting the SERVOn_FUNCTION to an RCINxx do this?

**Comment by Hwurzburg** [comment]
unneeded complexity....unneeded use of aux function places....a bit mask that designates if the passthru is scaled or not (current code) is all that is needed....you dont need to have any RC input assigned to any servo output...you can always arrange the ins and outs as needed currently, the only issue is that passthru does not scaling currently....

**Comment by timtuxworth** [comment]
> unneeded complexity....unneeded use of aux function places....a bit mask that designates if the passthru is scaled or not (current code) is all that is needed....you dont need to have any RC input assigned to any servo output...you can always arrange the ins and outs as needed currently, the only issue is that passthru does not scaling currently....

From this users perspective, it seems much simpler to just pick the RCINxx or RCINxx_MAPPED vs having to track down a separate parameter and figure out how to set it to match the SERVOn that I'm using. Are you talking code complexity? Because I'd suggest that the UX is much more complex with the bit mask.

**Comment by Hwurzburg** [comment]
yes code..we are removing functions weekly from the F4s....code space for generic functions is precious....and using up limited aux functions....only 255 are available for any future adds....and this can be accomplished without that
while making it a tiny bit more user friendly (although confusion between current passthru and this would STILL require reading the wiki to understand the differences) , its low bang per buck in my opinion....

**Comment by TunaLobster** [comment]
This approach is working on a Cube Orange now. Moved the RC min and max values around to test that the values were actually being mapped correctly.

Other approaches mentioned would probably be better than this one. I'll shelve the 16 new servo function approach in a different branch and try the bitmask approach.

**Comment by IamPete1** [comment]
Fixes https://github.com/ArduPilot/ardupilot/issues/18379

**Comment by TunaLobster** [comment]
Bitmask method is most of the way there. I need to figure out if the bitmask needs to be sent to the IOMCU. Current code problem is the bitmask parameter is not showing up in the GCS. I didn't think I was doing anything wildly different than the GPIO mask parameter.

**Comment by IamPete1** [comment]
I prefer the new aux functions, the param is a `int16` so were not going to run out. There maybe some transport cases where were using `uint8`, but were still a long way from 255 and we can fix that when we hit it. Better to only have one param, the servo function rather than to have to also go and set a bitmask.

**Comment by WickedShell** [comment]
> I prefer the new aux functions, the param is a `int16` so were not going to run out. There maybe some transport cases where were using `uint8`, but were still a long way from 255 and we can fix that when we hit it. Better to only have one param, the servo function rather than to have to also go and set a bitmask.

I'd strongly agree, and it's also just much more obvious what is happening as a user that way. It also helps limit the number of parameters we have to track and manage, which is valuable in it's own right.

**Comment by TunaLobster** [comment]
Branch is back to servo aux functions. Cleaned up and added the parameter metadata. This is going to need a wiki with a decent figure to help cover all the bases of information types. Marking for devcall as is and we'll talk about it then.

File: libraries/SRV_Channel/SRV_Channel.cpp [hunk 1]
```diff
@@ -64,10 +64,10 @@ const AP_Param::GroupInfo SRV_Channel::var_info[] = {
...
...SC,33:Motor1,34:Motor2,35:Motor3,36:Motor4,37:Motor5,38:Motor6,39:Motor7,40:Motor8,51:RCIN1,52:RCIN2,53:RCIN3,54:RCIN4,55:RCIN5,56:RCIN6,57:RCIN7,58:RCIN8,59:RCIN9,60:RCIN10,61:RCIN11,62:RCIN12,63:RCIN13,64:RCIN14,65:RCIN15,66:RCIN16,73:ThrottleLeft,74:ThrottleRight,75:TiltMotorFrontLeft,76:TiltMotorFrontRight,81:BoostThrottle,82:Motor9,83:Motor10,84:Motor11,85:Motor12,88:Winch,90:CameraISO,91:CameraAperture,92:CameraFocus,93:CameraShutterSpeed,94:Script1,95:Script2,96:Script3,97:Script4,98:Script5,99:Script6,100:Script7,101:Script8,102:Script9,103:Script10,104:Script11,105:Script12,106:Script13,107:Script14,108:Script15,109:Script16,120:NeoPixel1,121:NeoPixel2,122:NeoPixel3,123:NeoPixel4,124:RateRoll,125:RatePitch,126:RateThrust,127:RateYaw,129:ProfiLED1,130:ProfiLED2,131:ProfiLED3,132:ProfiLEDClock,133:Winch Clutch,134:SERVOn_MIN,135:SERVOn_TRIM,136:SERVOn_MAX,138:Alarm,139:Alarm Inverted,140:RCIN1_MAPPED,141:RCIN2_MAPPED,142:RCIN3_MAPPED,143:RCIN4_MAPPED,144:RCIN5_MAPPED,145:RCIN6_MAPPED,146:RCIN7_MAPPED,147:RCIN8_MAPPED,148:RCIN9_MAPPED,149:RCIN10_MAPPED,150:RCIN11_MAPPED,151:RCIN12_MAPPED,152:RCIN13_MAPPED,153:RCIN14_MAPPED,154:RCIN15_MAPPED,155:RCIN16_MAPPED,156:RCIN17_MAPPED
```
Comment (Hwurzburg) [inline libraries/SRV_Channel/SRV_Channel.cpp:69]: although its precedent for passthru, the mix of camel and all caps is distracting....but it should at least be a bit more explicative.....RCIN1Scaled,etc... MAPPED has no meaning...

**Review by Hwurzburg — CHANGES_REQUESTED** [review]

File: libraries/SRV_Channel/SRV_Channel.cpp (hunk 1 above)
Comment (timtuxworth) [reply libraries/SRV_Channel/SRV_Channel.cpp:69]: I agree with @Hwurzburg  about mapped vs scaled. It is actually scaled not mapped. On case - OMG NOT MORE CAPS PLEASE! :-)

**Review by IamPete1 — COMMENTED** [review]
I would vote to not bother with the IOMCU mixer stuff at all. Its increasingly unlikely that that functionality will ever save a vehicle.

**Comment by timtuxworth** [comment]
> I would vote to not bother with the IOMCU mixer stuff at all. Its increasingly unlikely that that functionality will ever save a vehicle.

From someone who is still reasonably new to this, I have generally understood that the message is "don't do anything in your transmitter - let ArduPilot manage everything", not having this scaling makes passthroughs an exception - except for passthroughs, for passthroughs you need to set trims and weights and offsets in your transmitter. Is it a good idea to have an exception?

**Comment by rmackay9** [comment]
As discussed on the dev call I agree with @WickedShell (and others) that just adding to the Enum (as is done in this PR) is a better approach.

File: libraries/SRV_Channel/SRV_Channel_aux.cpp [hunk 2]
```diff
@@ -32,6 +32,7 @@ void SRV_Channel::output_ch(void)
 {
 #ifndef HAL_BUILD_AP_PERIPH
     int8_t passthrough_from = -1;
+    bool passthrough_mapped;
```
Comment (tridge) [inline libraries/SRV_Channel/SRV_Channel_aux.cpp:35]: needs to init to false

**Review by tridge — CHANGES_REQUESTED** [review]
need to remove the IOMCU changes, test reversal at both ends, test zero DZ

File: libraries/SRV_Channel/SRV_Channel_aux.cpp (hunk 2 above)
Comment (TunaLobster) [reply libraries/SRV_Channel/SRV_Channel_aux.cpp]: Fixed

**Comment by TunaLobster** [comment]
Tested reversal, dead zone dramatically high, dead zone at zero, and trims at min. Log of the test is attached. Got to do a rebase with drops, fixup, and reword all at the same time!

[108 12-31-1969 6-00-00 PM.log](https://github.com/ArduPilot/ardupilot/files/9295952/108.12-31-1969.6-00-00.PM.log)

**Comment by IamPete1** [comment]
Some unexpected size changes, annoyingly for stuff that we don't save the elf diff for.

```
Binary Name      Text [B]        Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  --------------  -----------  -------------  ----------------------------  -------------------------
antennatracker   220 (+0.0293%)  0 (0.0000%)  4 (+0.0020%)   220 (+0.0293%)                                   280088
arducopter-heli  76 (+0.0075%)   0 (0.0000%)  4 (+0.0020%)   76 (+0.0075%)                                     18588
ardurover        76 (+0.0083%)   0 (0.0000%)  -4 (-0.0020%)  76 (+0.0083%)                                    112532
arducopter       76 (+0.0076%)   0 (0.0000%)  4 (+0.0020%)   76 (+0.0076%)                                     30612
ardusub          220 (+0.0239%)  0 (0.0000%)  4 (+0.0020%)   220 (+0.0239%)                                   111184
blimp            220 (+0.0311%)  0 (0.0000%)  -4 (-0.0020%)  220 (+0.0310%)                                   322572
arduplane        76 (+0.0074%)   0 (0.0000%)  -4 (-0.0020%)  76 (+0.0074%)                                       268
```

**Comment by TunaLobster** [comment]
I think I figured it out. The symbol for `RC_Channel::norm_input_dz() const` appeared in AntennaTracker and is 144 bytes. The other vehicles seem to already be using that method. So that accounts for the 144 byte difference. The board to board difference between vehicles is the same size but I don't know why that method was not already used.

**Review by IamPete1 — APPROVED** [review]
Looks good, Thanks!

**Review by tridge — APPROVED** [review]

File: libraries/SRV_Channel/SRV_Channel.cpp (hunk 1 above)
Comment (TunaLobster) [reply libraries/SRV_Channel/SRV_Channel.cpp]: Fixed and used the RCINxScaled

**Review by Hwurzburg — APPROVED** [review]
